### PR TITLE
refactor: begin refactor to jest while continuing to run tests in karma

### PR DESCRIPTION
**Pull request by ckifer**
We're moving to jest!

There's so much to move at once so only starting this refactor - not cleaning up much along the way for now.

* Add jest related packages to dev-deps
   * Upgrade React version to latest 16 (dev-dep)
* add combined test command to run both jest and karma
* move over some tests as started by @arcthur 
  *  some super minor cleanup

Also upgrade to latest TypeScript - this is needed to satisfy @types/jsdom

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,7 +29,9 @@
     "build-umd": "rimraf umd && cross-env NODE_ENV=production webpack --entry ./src/index.ts -o umd",
     "build-types": "rimraf types && npm run tsc",
     "demo": "webpack serve --config demo/webpack.config.js --port 3000 --host 127.0.0.1 --progress --profile --static demo/",
-    "test": "cross-env NODE_ENV=test karma start test/karma.conf.js",
+    "test": "npm run test-karma",
```
Comment (nikolasrieble) [inline package.json:32]: How about running both tests in CI?

File: package.json [hunk 2]
```diff
@@ -124,15 +130,15 @@
...
+    "react": "^16.14.0",
+    "react-dom": "^16.14.0",
     "react-router-dom": "^5.1.2",
     "react-test-renderer": "^16.12.0",
     "rimraf": "^3.0.2",
     "sinon": "^9.2.2",
     "terser-webpack-plugin": "^5.1.1",
     "ts-loader": "^8.0.12",
-    "typescript": "^4.1.3",
+    "typescript": "^4.9.4",
```
Comment (nikolasrieble) [inline package.json:141]: It might be nicer to do this upgrade in a separate PR, to reduce the blast radius of this PR.

Comment (nikolasrieble) [reply package.json:141]: As far as I understand, we do not need to upgrade typescript for jest, right?

File: package.json (hunk 1 above)
Comment (ckifer) [reply package.json:32]: Oops, that was the intent. What I get for making a PR tired 😄

File: package.json (hunk 2 above)
Comment (ckifer) [reply package.json:141]: I agree, but there are some unfortunate errors when installing jest-environment-jsdom that fail without a newer TS version (not sure what the exact version is).

Might have to use an older version of jest in general to make everything work. Will try that.

**Review by nikolasrieble — APPROVED** [review]
Nice, thank you!

**Review by arcthur — APPROVED** [review]

**Comment by arcthur** [comment]
the branch have conflict. pls solve the package info.

**Comment by ckifer** [comment]
Thanks, will fix!

**Comment by ckifer** [comment]
@arcthur @nikolasrieble upon merging I was getting some fun ts compilation errors because `@types/node` got upgraded. See failed build https://github.com/recharts/recharts/actions/runs/3678847245/jobs/6222592640. 

I pinned the `@types/node` version to 14 and addressed issues that were still yelling.

Changes since approval tried to be minimal. We'll need to go back and address these, especially the `@ts-ignore`. 2 more commits to fix prettier errors and a randomly failing test. Maybe `lint-staged` should be setup to autoformat?

**Comment by nikolasrieble** [comment]
> @arcthur @nikolasrieble upon merging I was getting some fun ts compilation errors because `@types/node` got upgraded. See failed build https://github.com/recharts/recharts/actions/runs/3678847245/jobs/6222592640.
> 
> I pinned the `@types/node` version to 14 and addressed issues that were still yelling.
> 
> Changes since approval tried to be minimal. We'll need to go back and address these, especially the `@ts-ignore`. 2 more commits to fix prettier errors and a randomly failing test. Maybe `lint-staged` should be setup to autoformat?

To include auto-formatting is a great idea. Makes everyones life easier.
